src/python/pants/backend/python/subsystems/pex_build_util.py

```diff
@@ -30,6 +34,86 @@
 from pants.util.contextutil import temporary_file
 
 
+# This bootstrapping technique for lazily loading requirements comes from the pex project test case:
+# https://github.com/pantsbuild/pex/commit/66aa5ef69e8b65004a548b44edc0b81dfa8ec736#diff-b5b57c9d4fc663f59ecaedb851c13bd8R23
+_IPEX_PREAMBLE = """\
+import json
+import os
+import sys
+
+from pex import resolver, vendor
+from pex.common import open_zip
+from pex.fetcher import Fetcher, PyPIFetcher
+from pex.interpreter import PythonInterpreter
+from pex.pex_builder import BOOTSTRAP_DIR, PEXBuilder
+from pex.pex_info import PexInfo
+from pex.platforms import Platform
+from pex.third_party.pkg_resources import DefaultProvider, ZipProvider, get_provider
+from pex.util import CacheHelper
+from pex.variables import ENV
+
+
+self = sys.argv[0]
+filename_base, ext = tuple(os.path.splitext(self))
+if ext == '.pex':
+  raise ValueError("Could not hydrate ipex: {}.pex must not be named with the .pex extension.".format(filename_base))
+hydrated_pex_file = '{}.pex'.format(filename_base)
```

Comment:
I'd reconsider this. The hydrated pex is not a regular pex as the example showed. If you read through that test, in the last lines it nukes the local ~/.pex cache which destroys the functioning of the "hydrated" pex. That's why I chose .ipex as the suffix for the hydrated pex and .ptex for the suffix of the pex building robot that filled in the gaps (ski repair reference).